// @aws-cdk/aws-sns-subscriptions/lib/email.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailSubscription = void 0;
const sns = require("@aws-cdk/aws-sns");
/**
 * Use an email address as a subscription target.
 *
 * Email subscriptions require confirmation.
 *
 * @stability stable
 */
class EmailSubscription {
    /**
     * @stability stable
     */
    constructor(emailAddress, props = {}) {
        this.emailAddress = emailAddress;
        this.props = props;
    }
    /**
     * Returns a configuration for an email address to subscribe to an SNS topic.
     *
     * @stability stable
     */
    bind(_topic) {
        return {
            subscriberId: this.emailAddress,
            endpoint: this.emailAddress,
            protocol: this.props.json ? sns.SubscriptionProtocol.EMAIL_JSON : sns.SubscriptionProtocol.EMAIL,
            filterPolicy: this.props.filterPolicy,
            deadLetterQueue: this.props.deadLetterQueue,
        };
    }
}
exports.EmailSubscription = EmailSubscription;
